### PR TITLE
feat: Initial commit

### PR DESCRIPTION
**Pull request by latonz**

### CODE REVIEW

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+jobs:
+  test:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
+      - uses: actions/setup-dotnet@v1
+      - run: dotnet tool restore
+      - run: dotnet build /p:TreatWarningsAsErrors=true
+      - run: dotnet test --no-build
+      - run: dotnet format --no-restore --verify-no-changes
```
[the remark this answers is not in the document]
Comment (latonz) [reply .github/workflows/test.yml:18]: Switched order.

File: README.md [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+// mapper declaration
+[Mapper]
+public interface IDtoMapper
+{
+    CarDto CarToCarDto(Car car);
+}
+
+// mapper usage
+var car = new Car { NumberOfSeats = 10, ... };
+var dto = DtoMapperImpl.Instance.CarToCarDto();
```
[the remark this answers is not in the document]
Comment (latonz) [reply README.md:33]: Jap, fixed.

File: src/Riok.MapSourceGen.Abstractions/MapperIgnoreAttribute.cs [hunk 3]
```diff
@@ -0,0 +1,23 @@
+namespace Riok.MapSourceGen.Abstractions;
+
+/// <summary>
+/// Ignores a property from the mapping.
+/// The use of `nameof()` is encuraged.
```
[the remark this answers is not in the document]
Comment (latonz) [reply src/Riok.MapSourceGen.Abstractions/MapperIgnoreAttribute.cs]: Fixed.

File: src/Riok.MapSourceGen.Abstractions/MapPropertyAttribute.cs [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+/// <summary>
+/// Specifies options for a property mapping.
+/// </summary>
+[AttributeUsage(AttributeTargets.Method, AllowMultiple = true)]
+public sealed class MapPropertyAttribute : Attribute
+{
+    /// <summary>
+    /// Maps a specified source property to the specified target property.
+    /// </summary>
+    /// <param name="source">The name of the source property. The use of `nameof()` is encuraged.</param>
```
[the remark this answers is not in the document]
Comment (latonz) [reply src/Riok.MapSourceGen.Abstractions/MapPropertyAttribute.cs]: Improved.

File: src/Directory.Build.props [hunk 5]
```diff
@@ -0,0 +1,34 @@
+<Project>
+  <!-- import parent: https://docs.microsoft.com/en-us/visualstudio/msbuild/customize-your-build -->
+  <Import Project="$([MSBuild]::GetPathOfFileAbove('Directory.Build.props', '$(MSBuildThisFileDirectory)../'))" />
+
+  <PropertyGroup>
+    <TargetFramework>netstandard2.0</TargetFramework>
```
[the remark this answers is not in the document]
Comment (latonz) [reply src/Directory.Build.props:6]: Afaik all source generator have to target `netstandard2.0`. Therefore I don't see any benefit of multi targeting.

File: README.md [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+dto.NumberOfSeats.Should().Be(10);
+```
+
+### Supported Mappings
```
[the remark this answers is not in the document]
Comment (latonz) [reply README.md:37]: Removed, we should think about how we would like to provide the documentation anyway. Not sure if GH Wiki should be the way to go.

File: src/Riok.Mapperly/Descriptors/DescriptorBuilder.cs [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+        _mapperDescriptor.InstanceName = mapperAttribute.InstanceName;
+
+        _defaultConfigurations.Add(typeof(MapEnumAttribute), new MapEnumAttribute(mapperAttribute.EnumMappingStrategy));
+    }
+
+    private string BuildName()
+    {
+        return !_mapperDescriptor.IsAbstractClassDefinition && _mapperSymbol.Name.StartsWith("I")
+            ? _mapperSymbol.Name.Substring(1)
+            : _mapperSymbol.Name + "Impl";
```
[the remark this answers is not in the document]
Comment (latonz) [reply src/Riok.Mapperly/Descriptors/DescriptorBuilder.cs:86]: Extracted.

File: src/Riok.Mapperly/Descriptors/MapperDescriptor.cs [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+        Name = baseName + "Impl";
+    }
+
+    public string? Namespace { get; set; }
+
+    public string Name { get; set; }
+
+    public string BaseName { get; }
+
+    public string FileName => $"{Name}.g.cs";
```
[the remark this answers is not in the document]
Comment (latonz) [reply src/Riok.Mapperly/Descriptors/MapperDescriptor.cs:22]: Extracted.

File: src/Riok.Mapperly/Descriptors/TypeMappings/ForEachAddDictionaryMapping.cs [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+            yield return CreateInstance(TargetVariableName, _typeToInstantiate, MemberAccess(source, CountPropertyName));
+        }
+        else
+        {
+            yield return CreateInstance(TargetVariableName, _typeToInstantiate);
+        }
+
+        var addMethod = MemberAccess(TargetVariableName, AddMethodName);
+        yield return ForEachStatement(
+            IdentifierName("var"),
```
[the remark this answers is not in the document]
Comment (latonz) [reply src/Riok.Mapperly/Descriptors/TypeMappings/ForEachAddDictionaryMapping.cs:57]: Fixed.

File: src/Riok.Mapperly/Descriptors/TypeMappings/TypeMapping.cs [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+
+    public ITypeSymbol SourceType { get; }
+
+    public ITypeSymbol TargetType { get; }
+
+    /// <summary>
+    /// Gets a value indicating if this mapping can be called / built by another mapping.
+    /// This should be <c>true</c> for most mappings.
+    /// </summary>
+    public virtual bool CanBeCalledByOtherMappings => true;
```
[the remark this answers is not in the document]
Comment (latonz) [reply src/Riok.Mapperly/Descriptors/TypeMappings/TypeMapping.cs:27]: Renamed.

File: src/Riok.Mapperly/Emit/SourceEmitter.cs [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+        var compilationUnit = CompilationUnit()
+            .WithMembers(SingletonList(WrapInNamespaceIfNeeded(descriptor.Namespace, classDeclaration)))
+            .WithLeadingTrivia(Nullable(true))
+            .NormalizeWhitespace();
+        return compilationUnit.ToFullString();
+    }
+
+    private static IEnumerable<MemberDeclarationSyntax> BuildMembers(MapperDescriptor descriptor)
+    {
+        // public static readonly MyMapper Instance = new MyMapperImpl();
```
[the remark this answers is not in the document]
Comment (latonz) [reply src/Riok.Mapperly/Emit/SourceEmitter.cs:27]: Depending on the mapper declaration (if it is an interface its only `MyMapper`, if it is an abstract class it is `MyMapperImpl`).

File: src/Riok.Mapperly/Riok.Mapperly.csproj [hunk 12]
```diff
@@ -0,0 +1,34 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <IncludeBuildOutput>false</IncludeBuildOutput>
+    <PackageReadmeFile>README.md</PackageReadmeFile>
+    <PublishRepositoryUrl>true</PublishRepositoryUrl>
+    <PackageId>Riok.Mapperly</PackageId>
+    <PackageLicense>Apache-2.0</PackageLicense>
```
[the remark this answers is not in the document]
Comment (latonz) [reply src/Riok.Mapperly/Riok.Mapperly.csproj:8]: As discussed, LICENSE file added.

File: README.md [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+| Enum             | Cast or by name mappings for enums, see [enum mapping configuration](#Enum)                     | From enum or any enum base type to an enum or any enum base type                                                                   |
+| Explicit cast    | Assign via an explicit cast                                                                     | Any to any if an explicit cast is defined                                                                                          |
+| ToString         | Convert via `ToString()` method call                                                            | Any to `string`                                                                                                                    |
+| Properties       | Creates a new target instance and maps each property                                            | Any to any, target type needs parameterless visible constructor                                                                    |
+
```
[the remark this answers is not in the document]
Comment (latonz) [reply README.md:55]: I'll do this in another PR. Maybe the integration test project could be reused?

**Review by CommonGuy — APPROVED** [review]

**Pull request merged**